feat(Community Permissions): Add search option in assets / collectibles dropdown

micieslak opened the pull request:
### What does the PR do

Adds/fixes search bar for panels in `HoldingsDropdown` according to the design.

Closes: #9042

<!-- Fill in the relevant information below to help us evaluate your proposed changes. -->

### Affected areas
Various components creating `HoldingsDropdown`.

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

[Kazam_screencast_00109.webm](https://user-images.githubusercontent.com/20650004/214864079-f449b43a-5312-4320-ab15-2a06bc44b48d.webm)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 75918ccb | [#1](https://ci.status.im/job/status-desktop/job/branches/job/linux-cpp/job/PR-9310/1/) | 2023-01-26 14:46:13 | ~6 min | `linux-cpp` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/linux-cpp/job/PR-9310/1//consoleText) |

caybro reviewed (approved):
LGTM

alexjba reviewed (commented):
It works fine in StoryBook, but in the app I see a different behaviour (please see the video below):
1. The search does not find items
2. The section "Search result" is not displayed after first char input
3. There are lots of warnings:
```
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Anniversary (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Anniversary (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Anniversary (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81:9: Unable to assign [undefined] to QString (qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/CryptoKitties (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/CryptoKitties (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/CryptoKitties (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81:9: Unable to assign QQmlListModel to QString (qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/SuperRare (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/SuperRare (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/SuperRare (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81:9: Unable to assign [undefined] to QString (qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Custom (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Custom (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Custom (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
Warning: qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81:9: Unable to assign [undefined] to QString (qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81, )
```
https://user-images.githubusercontent.com/47811206/215047008-f03e3f0c-4ba0-49bb-9800-1bd0ca4aac59.mov

noeliaSD reviewed (changes requested):
I've tested it in `storybook` and it works fine but when testing in the app this is the result:
1. Searcher doesn't work correcty in general.
2. In collectibles LIST MODE subitems view, the text `Search result` doesn't appear when I introduce the character `a`.
3. In collectibles THUMBNAILS MODE subitems view, the text `Search result` doesn't appear.

https://user-images.githubusercontent.com/97019400/215048287-86e1608e-7939-4e8b-a439-f6fa6091bc9e.mov

**Reflection / question:**

Does it make sense that the `searcher` is focused by default when the user navigates to the corresponding menu?

noeliaSD commented:
> It works fine in StoryBook, but in the app I see a different behaviour (please see the video below):
> 
> 1. The search does not find items
> 2. The section "Search result" is not displayed after first char input
> 3. There are lots of warnings:
> 
> ```
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Anniversary (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Anniversary (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Anniversary (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81:9: Unable to assign [undefined] to QString (qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/CryptoKitties (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/CryptoKitties (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/CryptoKitties (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81:9: Unable to assign QQmlListModel to QString (qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/SuperRare (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/SuperRare (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/SuperRare (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81:9: Unable to assign [undefined] to QString (qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Custom (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Custom (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/app/AppLayouts/Chat/controls/community/Custom (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
> Warning: qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81:9: Unable to assign [undefined] to QString (qrc:/app/AppLayouts/Chat/controls/community/ListDropdownContent.qml:81, )
> ```
> 
>  Screen.Recording.2023-01-27.at.10.54.58.mov

Oh!! You've been faster pushing your comments!! :)

noeliaSD commented on ui/app/AppLayouts/Chat/controls/community/ExtendedDropdownContent.qml:
```diff
@@ -30,38 +31,51 @@ Item {
         d.currentItemName = itemName
         d.currentItemSource = itemSource
-        root.state = d.listView_depth2_State
+        root.state = d.useThumbnailsOnDepth2
```
Thanks for fixing that!! Nice to keep the thumbnails option / view if it has been previously selected! :)

micieslak commented:
Thanks @alexjba and @noeliaSD ! It was quite nasty bug in `SortFilterProxyModel`, I provided workaround for that. Should work nice now. Please take a look once again :)

caybro commented:
> Thanks @alexjba and @noeliaSD ! It was quite nasty bug in `SortFilterProxyModel`, I provided workaround for that. Should work nice now. Please take a look once again :)

Would be worth to actually fix the bug upstream and update our copy ;)

micieslak commented:
> > Thanks @alexjba and @noeliaSD ! It was quite nasty bug in `SortFilterProxyModel`, I provided workaround for that. Should work nice now. Please take a look once again :)
> 
> Would be worth to actually fix the bug upstream and update our copy ;)

Yeah, that's my plan :)

noeliaSD commented:
Tested and it works correct now! Thanks for the fixes!

As discussed offline there are already 2 points to be solved that can also be addressed separately:
1. Searcher focus when openning or changing dropdown page.
2. Thumbnails view (if previously selected) after going back from selection view.

If you don't want to include these in this pr, please add new task for them! :)

micieslak commented:
Thanks @noeliaSD for testing and review,

regarding your points:

> 1. Searcher focus when openning or changing dropdown page.

Task for that: https://github.com/status-im/status-desktop/issues/9344

> 2. Thumbnails view (if previously selected) after going back from selection view.

Covered using `Settings`

Additionally I took into account `short` name in token's search as we agreed offline.
I also created a ticket for bug in `SortFilterProxyModel` for which the workaround has been created: https://github.com/status-im/status-desktop/issues/9345

alexjba reviewed (approved):
Very nice! 👍 Thank you!

anastasiyaig commented:
hey @micieslak , if u have time, could u please address some questions i do have for now, please take a look:

1. Search tabs are having titles in single, when designs show plural (Asset**s**, Collectible**s**)
2. How many assets we show by default without searching? There is a scroll bar appeared when there are 4 items under **Listed assets**, perhaps there is a possibility to remove the scroll when it is not necessary (expand the drop down list area to include some certain number of assets and display them without scrolling)? I am on laptop with 13 inches display
3. Community name does not reflect the actual name of the community (perhaps because of backend is not ready , but please confirm)
4. `All listed assets` link seems to be not implemented yet?

<img width="675" alt="Screenshot 2023-01-31 at 12 16 59" src="https://user-images.githubusercontent.com/82375995/215720501-b08642b7-98ef-4cc2-a15e-baf6f08fd7a5.png">

5. No placeholder when search is not successful 

<img width="675" alt="Screenshot 2023-01-31 at 12 23 04" src="https://user-images.githubusercontent.com/82375995/215720561-8e7b9bac-bac1-41d1-9a0e-f315f91f6d9f.png">


7. When entering a search criteria in Assets tab and switching to Collectible -> the search criteria is present in input box (i dont expect it to be like that but perhaps it is design, please clarify)


https://user-images.githubusercontent.com/82375995/215720808-2154172e-e86c-44dc-bea3-cd10e8020088.mov

8. Placeholder text for search is `search`, on designs it is `Search assets` and `Search collectibles`

<img width="675" alt="Screenshot 2023-01-31 at 12 27 09" src="https://user-images.githubusercontent.com/82375995/215721368-c93947f1-8109-466b-a645-14ea28ffa7e6.png">

micieslak commented:
Thank you @anastasiyaig for in-dept tests!

> 1. Search tabs are having titles in single, when designs show plural (Asset**s**, Collectible**s**)

Fixed

> 2. How many assets we show by default without searching? There is a scroll bar appeared when there are 4 items under **Listed assets**, perhaps there is a possibility to remove the scroll when it is not necessary (expand the drop down list area to include some certain number of assets and display them without scrolling)? I am on laptop with 13 inches display

It's not clear from the design for me. Some time ago I was pretty sure that by design the height is fixed. But after latest changes in the design I'm not sure about that. I think we should ask design team about that.

> 3. Community name does not reflect the actual name of the community (perhaps because of backend is not ready , but please confirm)

Yes, it's a matter of the backend which is not ready.

> 4. `All listed assets` link seems to be not implemented yet?

Yes, we have a separate ticket for that.

> 5. No placeholder when search is not successful

Do you know if we have a design for that? I couldn't find that. If it's not present, should be added ofc.

> 7. When entering a search criteria in Assets tab and switching to Collectible -> the search criteria is present in input box (i dont expect it to be like that but perhaps it is design, please clarify)

Yeah, for me also it's more natural behavior to clear search box when changing tabs. Fixed.
 
> 8. Placeholder text for search is `search`, on designs it is `Search assets` and `Search collectibles`

Fixed.

anastasiyaig commented:
@xAlisher could u please provide designs for the following states of Asset / Collectible drop down:

- No placeholder when search is not successful (nothing found based on search criteria)
![image](https://user-images.githubusercontent.com/82375995/215974237-00723cfd-1328-45de-9290-a59a339eb703.png)

- How many assets we show by default without searching? There is a scroll bar appeared when there are 4 items under Listed assets, perhaps there is a possibility to remove the scroll when it is not necessary (expand the drop down list area to include some certain number of assets and display them without scrolling)? I am on laptop with 13 inches display

xAlisher commented:
> @xAlisher could u please provide designs for the following states of Asset / Collectible drop down:
> 
> * No placeholder when search is not successful (nothing found based on search criteria)

![image](https://user-images.githubusercontent.com/7700914/215980579-9fbbea1a-32d8-456a-91ca-9b1476068a0b.png)
https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?node-id=25475%3A565830&t=YQyZLMyWtyuHRb8G-1

> * How many assets we show by default without searching? There is a scroll bar appeared when there are 4 items under Listed assets, perhaps there is a possibility to remove the scroll when it is not necessary (expand the drop down list area to include some certain number of assets and display them without scrolling)? I am on laptop with 13 inches display

![image](https://user-images.githubusercontent.com/7700914/215980739-ccbed090-806d-4a15-ab8a-4e40cc45dd10.png)
https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?node-id=25475%3A565859&t=YQyZLMyWtyuHRb8G-1

anastasiyaig commented:
@micieslak could u please implement these 2 screens that Alisher provided? 🙏

noeliaSD commented:
> > * How many assets we show by default without searching? There is a scroll bar appeared when there are 4 items under Listed assets, perhaps there is a possibility to remove the scroll when it is not necessary (expand the drop down list area to include some certain number of assets and display them without scrolling)? I am on laptop with 13 inches display
> 
> ![image](https://user-images.githubusercontent.com/7700914/215980739-ccbed090-806d-4a15-ab8a-4e40cc45dd10.png) https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?node-id=25475%3A565859&t=YQyZLMyWtyuHRb8G-1

@micieslak could you add a new ticked for the calculation of the dropdown heigh? Dynamic, adapting the heigh to the elements shown but with a maximum heigh like in this design!

micieslak commented:
@anastasiyaig @noeliaSD 

The first issue is fixed (results placeholder).
For the second one (dynamic height) there is a new ticket: https://github.com/status-im/status-desktop/issues/9385

Thank you for your input!

anastasiyaig commented:
thank you @micieslak it is tested
i also created  a squish task to cover that someday https://github.com/status-im/status-desktop/issues/9390
